### DISCUSSION TITLE
U

### DISCUSSION THREAD

**Discussion by 0x-codeX** [Cyfrin/foundry-full-course-cu#1629] [Q&A] [answered]
I have tried to authorize owner, and all other tests passed except this.

`
   function testRevertsIfMintFails() public {
        // Arrange - Setup
        MockFailedMintDSC mockDsc = new MockFailedMintDSC();
        tokenAddresses = [weth];
        feedAddresses = [ethUsdPriceFeed];
        address owner = msg.sender;
        vm.prank(USER);
        DSCEngine mockDsce = new DSCEngine(tokenAddresses, feedAddresses, address(mockDsc));
        mockDsc.transferOwnership(address(mockDsce));
        // Arrange - User
        vm.startPrank(owner);
        ERC20Mock(weth).approve(address(mockDsce), AMOUNT_COLLATERAL);
        vm.expectRevert(DSCEngine.DSCEngine__MintFailed.selector);
        mockDsce.depositCollateralAndMintDsc(weth, AMOUNT_COLLATERAL, amountToMint);
        vm.stopPrank();
    }
`

The error

`
Ran 1 test for test/unit/DSCEngineTest.t.sol:DSCEngineTest
[FAIL. Reason: OwnableUnauthorizedAccount(0x34A1D3fff3958843C43aD80F30b94c510645C316)] testRevertsIfMintFails() (gas: 1755663)
Traces:
  [9331228] DSCEngineTest::setUp()
    ├─ [4152257] → new DeployDSC@0xFEfC6BAF87cF3684058D62Da40Ff3A795946Ab06
    │   └─ ← 20676 bytes of code
    ├─ [4882090] DeployDSC::run()
    │   ├─ [3007844] → new HelperConfig@0x6d2eed85750d316088343D6d5e91ca59eb052768
    │   │   ├─ [0] VM::startBroadcast()
    │   │   │   └─ ← ()
    │   │   ├─ [372255] → new MockV3Aggregator@0xA8452Ec99ce0C64f20701dB7dD3abDb607c00496
    │   │   │   └─ ← 1416 bytes of code
    │   │   ├─ [449186] → new ERC20Mock@0xBb2180ebd78ce97360503434eD37fcf4a1Df61c3
    │   │   │   └─ ← 2499 bytes of code
    │   │   ├─ [372255] → new MockV3Aggregator@0xDB8cFf278adCCF9E9b5da745B44E754fC4EE3C76
    │   │   │   └─ ← 1416 bytes of code
    │   │   ├─ [449186] → new ERC20Mock@0x50EEf481cae4250d252Ae577A09bF514f224C6C4
    │   │   │   └─ ← 2499 bytes of code
    │   │   ├─ [0] VM::stopBroadcast()
    │   │   │   └─ ← ()
    │   │   └─ ← 10462 bytes of code
    │   ├─ [912] HelperConfig::activeNetworkConfig() [staticcall]
    │   │   └─ ← MockV3Aggregator: [0xA8452Ec99ce0C64f20701dB7dD3abDb607c00496], MockV3Aggregator: [0xDB8cFf278adCCF9E9b5da745B44E754fC4EE3C76], ERC20Mock: [0xBb2180ebd78ce97360503434eD37fcf4a1Df61c3], ERC20Mock: [0x50EEf481cae4250d252Ae577A09bF514f224C6C4], 77814517325470205911140941194401928579557062014761831930645393041380819009408 [7.781e76]
    │   ├─ [0] VM::startBroadcast(<pk>)
    │   │   └─ ← ()
    │   ├─ [590846] → new DecentralizedStableCoin@0x5FbDB2315678afecb367f032d93F642f64180aa3
    │   │   ├─ emit OwnershipTransferred(previousOwner: 0x0000000000000000000000000000000000000000, newOwner: 0xf39Fd6e51aad88F6F4ce6aB8827279cffFb92266)
    │   │   └─ ← 3258 bytes of code
    │   ├─ [1043371] → new DSCEngine@0xe7f1725E7734CE288F8367e1Bb143E90bb3F0512
    │   │   └─ ← 5640 bytes of code
    │   ├─ [2421] DecentralizedStableCoin::transferOwnership(DSCEngine: [0xe7f1725E7734CE288F8367e1Bb143E90bb3F0512])
    │   │   ├─ emit OwnershipTransferred(previousOwner: 0xf39Fd6e51aad88F6F4ce6aB8827279cffFb92266, newOwner: DSCEngine: [0xe7f1725E7734CE288F8367e1Bb143E90bb3F0512])
    │   │   └─ ← ()
    │   ├─ [0] VM::stopBroadcast()
    │   │   └─ ← ()
    │   └─ ← DecentralizedStableCoin: [0x5FbDB2315678afecb367f032d93F642f64180aa3], DSCEngine: [0xe7f1725E7734CE288F8367e1Bb143E90bb3F0512], HelperConfig: [0x6d2eed85750d316088343D6d5e91ca59eb052768]
    ├─ [912] HelperConfig::activeNetworkConfig() [staticcall]
    │   └─ ← MockV3Aggregator: [0xA8452Ec99ce0C64f20701dB7dD3abDb607c00496], MockV3Aggregator: [0xDB8cFf278adCCF9E9b5da745B44E754fC4EE3C76], ERC20Mock: [0xBb2180ebd78ce97360503434eD37fcf4a1Df61c3], ERC20Mock: [0x50EEf481cae4250d252Ae577A09bF514f224C6C4], 77814517325470205911140941194401928579557062014761831930645393041380819009408 [7.781e76]
    ├─ [0] VM::deal(0x0000000000000000000000000000000000000001, 10000000000000000000 [1e19])
    │   └─ ← ()
    ├─ [46789] ERC20Mock::mint(0x0000000000000000000000000000000000000001, 10000000000000000000 [1e19])
    │   ├─ emit Transfer(from: 0x0000000000000000000000000000000000000000, to: 0x0000000000000000000000000000000000000001, value: 10000000000000000000 [1e19])
    │   └─ ← ()
    ├─ [46789] ERC20Mock::mint(0x0000000000000000000000000000000000000001, 10000000000000000000 [1e19])
    │   ├─ emit Transfer(from: 0x0000000000000000000000000000000000000000, to: 0x0000000000000000000000000000000000000001, value: 10000000000000000000 [1e19])
    │   └─ ← ()
    └─ ← ()

  [1755663] DSCEngineTest::testRevertsIfMintFails()
    ├─ [591266] → new MockFailedMintDSC@0x2a9e8fa175F45b235efDdD97d2727741EF4Eee63
    │   ├─ emit OwnershipTransferred(previousOwner: 0x0000000000000000000000000000000000000000, newOwner: 0xf39Fd6e51aad88F6F4ce6aB8827279cffFb92266)
    │   └─ ← 3297 bytes of code
    ├─ [0] VM::prank(user: [0x6CA6d1e2D5347Bfab1d91e883F1915560e09129D])
    │   └─ ← ()
    ├─ [998042] → new DSCEngine@0x7BD1119CEC127eeCDBa5DCA7d1Bd59986f6d7353
    │   └─ ← 5576 bytes of code
    ├─ [615] MockFailedMintDSC::transferOwnership(DSCEngine: [0x7BD1119CEC127eeCDBa5DCA7d1Bd59986f6d7353])
    │   └─ ← OwnableUnauthorizedAccount(0x34A1D3fff3958843C43aD80F30b94c510645C316)
    └─ ← OwnableUnauthorizedAccount(0x34A1D3fff3958843C43aD80F30b94c510645C316)
`

**Comment by EngrPips** [comment]
```javascript
function testRevertsIfMintFails() public {
// Arrange - Setup
MockFailedMintDSC mockDsc = new MockFailedMintDSC();
tokenAddresses = [weth];
feedAddresses = [ethUsdPriceFeed];
address owner = msg.sender;
vm.prank(USER);
DSCEngine mockDsce = new DSCEngine(tokenAddresses, feedAddresses, address(mockDsc));
mockDsc.transferOwnership(address(mockDsce));
// Arrange - User
vm.startPrank(owner);
ERC20Mock(weth).approve(address(mockDsce), AMOUNT_COLLATERAL);

    vm.expectRevert(DSCEngine.DSCEngine__MintFailed.selector);
    mockDsce.depositCollateralAndMintDsc(weth, AMOUNT_COLLATERAL, amountToMint);
    vm.stopPrank();
}
```

**Reply by 0x-codeX** [reply]
Thanks @EngrPips this is exactly what I wrote the function, word for word and it is not working.

**Reply by EngrPips** [reply]
Could you please open the `MockFailedMintDSC` and check where exactly the contract is instructed to revert with `OwnableUnauthorizedAccount(0x34A1D3fff3958843C43aD80F30b94c510645C316)`?  and also the below code should work fine

```javascript
function testRevertsIfMintFails() public {
// Arrange - Setup
MockFailedMintDSC mockDsc = new MockFailedMintDSC();
tokenAddresses = [weth];
feedAddresses = [ethUsdPriceFeed];
address owner = msg.sender;
vm.prank(USER);
DSCEngine mockDsce = new DSCEngine(tokenAddresses, feedAddresses, address(mockDsc));
vm.prank(mockDsc.owner())
mockDsc.transferOwnership(address(mockDsce));
// Arrange - User
vm.startPrank(owner);
ERC20Mock(weth).approve(address(mockDsce), AMOUNT_COLLATERAL);

    vm.expectRevert(DSCEngine.DSCEngine__MintFailed.selector);
    mockDsce.depositCollateralAndMintDsc(weth, AMOUNT_COLLATERAL, amountToMint);
    vm.stopPrank();
}
```

**Reply by 0x-codeX** [reply]
Below is what I found in Ownable.sol

`
 function _checkOwner() internal view virtual {
        if (owner() != _msgSender()) {
            revert OwnableUnauthorizedAccount(_msgSender());
        }
    }
`

**Reply by EngrPips** [reply]
```javascript
function testRevertsIfMintFails() public {
// Arrange - Setup
MockFailedMintDSC mockDsc = new MockFailedMintDSC();
tokenAddresses = [weth];
feedAddresses = [ethUsdPriceFeed];
address owner = msg.sender;
vm.prank(USER);
DSCEngine mockDsce = new DSCEngine(tokenAddresses, feedAddresses, address(mockDsc));
vm.prank(mockDsc.owner())
mockDsc.transferOwnership(address(mockDsce));
// Arrange - User
vm.startPrank(owner);
ERC20Mock(weth).approve(address(mockDsce), AMOUNT_COLLATERAL);

    vm.expectRevert(DSCEngine.DSCEngine__MintFailed.selector);
    mockDsce.depositCollateralAndMintDsc(weth, AMOUNT_COLLATERAL, amountToMint);
    vm.stopPrank();
}
```

Did you try the above code, it should pass

**Reply by 0x-codeX** [reply, marked as answer]
Thanks again @EngrPips I tried with the added line of code and it gave another but familiar error, fixed that as well, all good.

**Reply by EngrPips** [reply]
Awesome, I am glad you figured it out. Keep crushing it
